internal/pkg/skuba/deployments/ssh/kubelet.go

```diff
@@ -18,17 +18,156 @@
+	KubeletServerCertAndKeyBaseName = "kubelet"
+	// KubeletServerCertName defines kubelet server certificate name
+	KubeletServerCertName = "kubelet.crt"
+	// KubeletServerKeyName defines kubelet server key name
+	KubeletServerKeyName = "kubelet.key"
+)
+
+func kubeletCreateRootCert(t *Target, data interface{}) error {
+	caCertFilePath := filepath.Join(skubaconstants.PkiDir(), KubeletCACertName)
+	caKeyFilePath := filepath.Join(skubaconstants.PkiDir(), KubeletCAKeyName)
+	if canReadCertAndKey, _ := certutil.CanReadCertAndKey(caCertFilePath, caKeyFilePath); canReadCertAndKey {
+		klog.V(1).Info("kubelet root ca cert and key already exists")
+		return nil
+	}
+
+	cfg := &certutil.Config{
+		CommonName: "kubelet-ca",
+	}
+	caCert, caKey, err := pkiutil.NewCertificateAuthority(cfg)
+	if err != nil {
+		return errors.Wrap(err, "couldn't generate kubelet CA certificate")
+	}
+
+	if err := pkiutil.WriteCertAndKey(skubaconstants.PkiDir(), KubeletCACertAndKeyBaseName, caCert, caKey); err != nil {
+		return errors.Wrap(err, "failure while saving kubelet CA certificate and key")
+	}
+
+	return nil
+}
+
+func kubeletCreateServerCert(t *Target, data interface{}) error {
+	// Read kubelet root ca certificate and key
+	caCert, caKey, err := pkiutil.TryLoadCertAndKeyFromDisk(skubaconstants.PkiDir(), KubeletCACertAndKeyBaseName)
+	if err != nil {
+		return errors.Wrap(err, "failure loading kubelet CA certificate authority")
+	}
+
+	host := t.target.Nodename
+	altNames := certutil.AltNames{}
+	if ip := net.ParseIP(host); ip != nil {
+		altNames.IPs = append(altNames.IPs, ip)
+	} else {
+		altNames.DNSNames = append(altNames.DNSNames, host)
+	}
+
+	// Create AltNames with defaults DNSNames/IPs
+	stdout, _, err := t.silentSsh("hostname", "-I")
+	if err != nil {
+		return err
+	}
+	for _, addr := range strings.Split(stdout, " ") {
+		if ip := net.ParseIP(addr); ip != nil {
+			altNames.IPs = append(altNames.IPs, ip)
+		}
+	}
+
+	alternateIPs := []net.IP{net.ParseIP(t.target.Target), net.IPv4(127, 0, 0, 1), net.IPv6loopback}
+	alternateDNS := []string{"localhost"}
+
+	altNames.IPs = append(altNames.IPs, alternateIPs...)
+	altNames.DNSNames = append(altNames.DNSNames, alternateDNS...)
+
+	cfg := &certutil.Config{
+		CommonName: host,
+		AltNames:   altNames,
+		Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+	}
+	cert, key, err := pkiutil.NewCertAndKey(caCert, caKey, cfg)
+	if err != nil {
+		return errors.Wrap(err, "couldn't generate kubelet server certificate")
+	}
+
+	// Save kubelet server certificate and key to local temporarily
+	if err := pkiutil.WriteCertAndKey(skubaconstants.PkiDir(), host, cert, key); err != nil {
+		return errors.Wrapf(err, "failure while saving kubelet server %s certificate and key", host)
+	}
+
+	// Upload root ca cert
+	if err := t.target.UploadFile(filepath.Join(skubaconstants.PkiDir(), KubeletCACertName), filepath.Join(KubeletCertAndKeyDir, KubeletCACertName)); err != nil {
```

Comment:
I think we might want to split this "state" into two or three different ones: upload ca cert, generate new cert, upload generated cert and key.

Reply:
NP!